### PR TITLE
Add selector-pseudo-class-whitelist rule

### PR DESCRIPTION
**Pull request by jeddy3**
Ref #1119

It feels like it’s been yonks since I wrote a rule. There’s one or two proposals that I know I’d use on my next project, so I figured I'd try getting them done.

### CODE REVIEW

File: src/rules/selector-pseudo-class-whitelist/__tests__/index.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    message: messages.rejected("nth-LAST-child"),
+    line: 1,
+    column: 4,
+  }, {
+    code: "a,\n:global {}",
+    message: messages.rejected("global"),
+    line: 2,
+    column: 1,
+  }, {
+    code: "input:-ms-input-placeholder {}",
```
Comment (evilebottnawi) [inline src/rules/selector-pseudo-class-whitelist/__tests__/index.js]: Add test `input:-MS-INPUT-PLACEHOLDER {}` to tests case insensitivity for vendor properties

**Comment by evilebottnawi** [comment]
One comment. LGTM. :+1:

**Comment by ntwb** [comment]
LGTM

File: src/rules/selector-pseudo-class-whitelist/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+Specify a whitelist of allowed pseudo-class selectors.
+
+```css
+  a:hover {}
+/** ↑
+ * These pseudo-class selectors */
+```
+
+This rule ignores selectors that use variables interpolation e.g. `:#{$variable} {}`.
```
Comment (davidtheclark) [inline src/rules/selector-pseudo-class-whitelist/README.md]: variable (singular)

File: src/rules/selector-pseudo-class-whitelist/__tests__/index.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+testRule(rule, {
+  ruleName,
+  config: [[
+    "hover",
+    "nth-child",
+    "root",
+    "placeholder",
+  ]],
+  skipBasicChecks: true,
```
Comment (davidtheclark) [inline src/rules/selector-pseudo-class-whitelist/__tests__/index.js]: Curious: why are you skipping the basic checks on all of these? Is there a pseudo-class in them?

File: src/rules/selector-pseudo-class-whitelist/index.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+      if (!isStandardSyntaxSelector(selector)) { return }
+      if (selector.indexOf(":") === -1) { return }
+
+      parseSelector(selector, result, rule, selectorTree => {
+        selectorTree.walkPseudos(pseudoNode => {
+          const pseudo = pseudoNode.value
+
+          // Ignore pseudo-elements
+          if (pseudo.indexOf("::") !== -1) { return }
```
Comment (davidtheclark) [inline src/rules/selector-pseudo-class-whitelist/index.js]: Could you instead just check the beginning of the string? This, for example, will ignore `a:not(a::selection)`, mistakenly I think.

File: src/rules/selector-pseudo-class-whitelist/index.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+          const pseudoName = pseudo.replace(/:+/, "")
```
Comment (davidtheclark) [inline src/rules/selector-pseudo-class-whitelist/index.js]: Again, don't we only want to deal with the colons at the very beginning?

File: src/rules/selector-pseudo-class-whitelist/__tests__/index.js (hunk 3 above)
Comment (jeddy3) [inline src/rules/selector-pseudo-class-whitelist/__tests__/index.js]: Yep, `:global` -https://github.com/stylelint/stylelint/blob/master/src/testUtils/basicChecks.js#L15-L18

It surprised me too. Perhaps we should remove it from the basicChecks?

Comment (davidtheclark) [inline src/rules/selector-pseudo-class-whitelist/__tests__/index.js]: Nah, we put it there for a reason.

File: src/rules/selector-pseudo-class-whitelist/index.js (hunk 4 above)
Comment (jeddy3) [inline src/rules/selector-pseudo-class-whitelist/index.js]: Yes., good catch!

This will need to be tested and patched in the other `-pseudo-` rules too.

Comment (jeddy3) [inline src/rules/selector-pseudo-class-whitelist/index.js]: Actually, this is fine. The parser extracts and iterates through *singular* pseudos e.g.:

```js
:hover
:nth-child
:root
:HOVER
:hover
:nth-child
::before
:nth-child
::before
:-moz-placeholder
:-MOZ-PLACEholder
:focus
:nth-LAST-child
:global
:-ms-input-placeholder
:-Ms-INPUT-placeholder
:not
::selection
:nth-child
:nth-LAST-child
:hover
```

So, the colons will always be at the start. I've added a test for this.

Comment (davidtheclark) [inline src/rules/selector-pseudo-class-whitelist/index.js]: Is there any reason not to change this line to 

```js
if (pseudo.slice(0, 2) !== "::") { return }
```

which would not search the whole string when only the beginning matters?

File: src/rules/selector-pseudo-class-whitelist/index.js (hunk 5 above)
Comment (davidtheclark) [inline src/rules/selector-pseudo-class-whitelist/index.js]: Is there any reason not to change this to

```js
const pseudoName = pseudo.slice(2)
```

?

File: src/rules/selector-pseudo-class-whitelist/index.js (hunk 4 above)
Comment (jeddy3) [inline src/rules/selector-pseudo-class-whitelist/index.js]: > Is there any reason not to change this line

Nope. You're right that this line is better. The *intent* feels clearer.

Changing now.

**Comment by jeddy3** [comment]
Ready for review.

**Comment by evilebottnawi** [comment]
LGTM

**Comment by davidtheclark** [comment]
Great!

**Pull request merged**